crates/sui-core/src/validator_client_monitor/monitor.rs

```diff
@@ -268,12 +271,28 @@ impl<A: Clone> ValidatorClientMonitor<A> {
         // an out-of-date committee.
         let mut validator_with_latencies: Vec<_> = committee
             .names()
-            .map(|v| (*v, cached_latencies.get(v).cloned().unwrap_or(0.0)))
+            .map(|v| {
+                (
+                    *v,
+                    cached_latencies.get(v).cloned().unwrap_or(Duration::ZERO),
+                )
+            })
             .collect();
+        if validator_with_latencies.is_empty() {
+            return vec![];
+        }
         // Sort by latency in ascending order. We want to select the validators with the lowest latencies.
-        validator_with_latencies.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
-
-        let k = k.min(validator_with_latencies.len());
+        validator_with_latencies.sort_by_key(|(_, latency)| *latency);
+
+        // Shuffle validators within delta of the lowest latency, for load balancing.
+        let lowest_latency = validator_with_latencies[0].1;
+        let threshold = lowest_latency.mul_f64(1.0 + delta);
+        let k = validator_with_latencies
+            .iter()
+            .enumerate()
+            .find(|(_, (_, latency))| *latency > threshold)
+            .map(|(i, _)| i)
+            .unwrap_or(validator_with_latencies.len());
```

Comment:
I would still recommend having some sort of minimum here to avoid monopolising only a few validators in more extreme cases.

```suggestion
            .unwrap_or(validator_with_latencies.len())
            .max(MIN_K);
```

Reply:
Maybe also let's add a metric to track `k` here to get a better understanding how this is formed.

Reply:
I want to verify if this current approach is enough for load balancing through experiments. Even if a larger K is used, different full nodes may still end up using a single validators disproportionally, let alone malicious cases. Also, from metrics, latencies increase noticeably (10%) from 5%-tile to 25%-tile. I want to see if using 10% delta or even lower value can be good enough here.